File: src/providers/mock.ts
/**
 * Mock provider — produces deterministic-ish stats for wiring up the
 * plugin end-to-end before any real fetcher is built.
 *
 * The values drift slowly over time so the button fill visibly
 * animates during development.
 */

import type { MetricValue, Provider, ProviderSnapshot } from "./types.ts";

export class MockProvider implements Provider {
  readonly id = "mock";
  readonly name = "Mock";
  readonly brandColor = "#3b82f6"; // neutral dev-blue; not a CodexBar-branded hue
  readonly metricIds = [
    "session-percent",
    "weekly-percent",
    "credits",
  ] as const;

  async fetch(): Promise<ProviderSnapshot> {
    const now = Date.now();
    // 10-minute sine sweep so the button is obviously animating.
    const t = (now / 1000 / 60 / 10) * Math.PI * 2;
    const session = 50 + Math.sin(t) * 45;
    const weekly = 50 + Math.sin(t * 0.3) * 40;
    const credits = 250 + Math.cos(t) * 100;

    const metrics: MetricValue[] = [
      {
        id: "session-percent",
        label: "SESSION",
        name: "Session window remaining",
        value: Math.round(session),
        unit: "%",
        ratio: session / 100,
        direction: "up",
        resetInSeconds: 60 * 60 * 3,
        updatedAt: new Date(now),
      },
      {
        id: "weekly-percent",
        label: "WEEKLY",
        name: "Weekly window remaining",
        value: Math.round(weekly),
        unit: "%",
        ratio: weekly / 100,
        direction: "up",
        resetInSeconds: 60 * 60 * 24 * 4,
        updatedAt: new Date(now),
      },
      {
        id: "credits",
        label: "CREDITS",
        name: "Credits remaining",
        value: Math.round(credits),
        ratio: credits / 400,
        direction: "up",
        updatedAt: new Date(now),
      },
    ];

    return {
      providerId: this.id,
      providerName: this.name,
      source: "mock",
      metrics,
      status: "operational",
    };
  }
}

File: internal/providers/mock.go
package providers

import (
	"math"
	"time"
)

// MockProvider generates deterministic sine-wave data for development.
type MockProvider struct{}

// ID returns the provider identifier used by the registry.
func (MockProvider) ID() string { return "mock" }

// Name returns the human-readable provider name.
func (MockProvider) Name() string { return "Mock" }

// BrandColor returns the accent color used on button faces.
func (MockProvider) BrandColor() string { return "#3b82f6" }

// BrandBg returns the background color used on button faces.
func (MockProvider) BrandBg() string { return "#111827" }

// MetricIDs enumerates the metrics this provider emits.
func (MockProvider) MetricIDs() []string {
	return []string{"session-percent", "weekly-percent", "credits"}
}

// Fetch returns a synthetic Snapshot driven by time-based sine waves.
func (MockProvider) Fetch(_ FetchContext) (Snapshot, error) {
	t := float64(time.Now().UnixMilli()) / 1000.0

	sessionVal := 50 + 50*math.Sin(t/300)
	weeklyVal := 50 + 50*math.Sin(t/600)
	credits := 50 + 50*math.Cos(t/900)

	sessionRatio := sessionVal / 100
	weeklyRatio := weeklyVal / 100

	resetSec := 18000 - math.Mod(t, 18000) // 5h window

	sv := math.Round(sessionVal)
	wv := math.Round(weeklyVal)
	cv := math.Round(credits*100) / 100

	return Snapshot{
		ProviderID:   "mock",
		ProviderName: "Mock",
		Source:       "mock",
		Metrics: []MetricValue{
			{
				ID:            "session-percent",
				Label:         "SESSION",
				Name:          "Session % remaining",
				Value:         sv,
				NumericValue:  &sv,
				NumericUnit:   "percent",
				Unit:          "%",
				Ratio:         &sessionRatio,
				Direction:     "up",
				ResetInSeconds: &resetSec,
			},
			{
				ID:           "weekly-percent",
				Label:        "WEEKLY",
				Name:         "Weekly % remaining",
				Value:        wv,
				NumericValue: &wv,
				NumericUnit:  "percent",
				Unit:         "%",
				Ratio:        &weeklyRatio,
				Direction:    "up",
			},
			{
				ID:          "credits",
				Label:       "CREDITS",
				Name:        "Credits remaining",
				Value:       cv,
				NumericValue: &cv,
				NumericUnit: "count",
			},
		},
	}, nil
}

// init registers MockProvider with the package-wide registry on import.
func init() {
	Register(MockProvider{})
}
